### PR TITLE
Refactor Sequence Protocol

### PR DESCRIPTION
**Pull request by qingshi163**

### CODE REVIEW

**Comment by youknowone** [comment]
@Snowapril would you review this pr?

File: vm/src/builtins/type.rs [hunk 1]
```diff
@@ -194,7 +204,7 @@ impl PyType {
             None,
         );
 
-        for attr_name in new_type.attributes.read().keys() {
+        for attr_name in attr_name_set {
             if attr_name.as_str().starts_with("__") && attr_name.as_str().ends_with("__") {
```
Comment (youknowone) [inline vm/src/builtins/type.rs:208]: these check seems to be able to be done when inserting to attr_name_set.

**Review by youknowone — APPROVED** [review]

**Comment by Snowapril** [comment]
I also agree to put all sequence methods into function as static local. It'll reduce inessential memory residency. It seems good to me.

File: vm/src/protocol/mapping.rs [hunk 2]
```diff
@@ -45,7 +45,7 @@ impl PyMappingMethods {
         ];
 
         fn length(mapping: &PyMapping, vm: &VirtualMachine) -> PyResult<usize> {
-            crate::types::slot_length(mapping.obj, vm)
+            crate::types::len_wrapper(mapping.obj, vm)
```
Comment (Snowapril) [inline vm/src/protocol/mapping.rs:48]: Could this be called through the protocol interface? It seems it'll go through the mro resolution inside python side, doesn't it?

Comment (Snowapril) [reply vm/src/protocol/mapping.rs:48]: Oh is it for generic cases? Ok I see

**Pull request merged**